File: linkedlist/leetcode_Remove_Duplicates_from_Sorted_List.java
/*Remove_Duplicates_from_Sorted_List

Given a sorted linked list, delete all duplicates such that each element appear
only once.
For example,
Given 1->1->2, return 1->2.
Given 1->1->2->3->3, return 1->2->3.
*/

class leetcode_Remove_Duplicates_from_Sorted_List {
    public static void main(String[] args) {
        ListNode head = new ListNode(1);
        head.next = new ListNode(1);
        deleteDuplicates(head);
    }
    public static ListNode deleteDuplicates(ListNode head) {
        // The idea behind this solution is same as 
        // leetcode_Remove_Duplicates_from_Sorted_Array
        if (head == null) {
            return null;
        }
        ListNode cursor = head.next;
        ListNode processed = head;
        while (cursor != null) {
            if (cursor.val != processed.val) {
                processed.next.val = cursor.val;
                processed = processed.next;
            }
            cursor = cursor.next;
        }
        processed.next = null;
        return head;
    }
}


class ListNode {
    ListNode next;
    int val;
    ListNode(int x) {
        val = x;
        next = null;
    }
}

/* Python Version
def deleteDuplicates(self, head):
    if not head:
        return head

    processed = head
    cursor = head.next
    prev = head # Only need one pointer instead cursor and prev both, but having two is clearer.
    
    while cursor:
        if cursor.val != prev.val:
            processed.next.val = cursor.val
            processed = processed.next
        cursor = cursor.next
        prev = prev.next
    
	# Remove the rest    
    processed.next = None

    return head
*/

File: linkedlist/leetcode_Remove_Duplicates_from_Sorted_List.cpp
/**
 * Copyright 2016
 * author: madarfacar
 *
 * leetcode remove duplicates from sorted list
 */

#include <stdio.h>


/**
 * Definition for singly-linked list.
 */
struct ListNode {
  int val;
  ListNode *next;
  ListNode(int x) : val(x), next(NULL) {}

  void printAll() {
    printf("%d -> ", val);

    ListNode* cursor = next;
    while (cursor != NULL) {
      printf("%d -> ", cursor->val);
      cursor = cursor->next;
    }
    printf("NULL\n");
  }
};


class Solution {
 public:
  ListNode* deleteDuplicates(ListNode* head) {
    if (head == NULL) {
      return NULL;
    }
    ListNode* cursor = head->next;
    ListNode* dedup_cursor = head;

    while (cursor != NULL) {
      if (cursor->val != dedup_cursor->val)  {
        dedup_cursor->next->val = cursor->val;
        dedup_cursor = dedup_cursor->next;
      }
      cursor = cursor->next;
    }
    // Unlink.
    dedup_cursor->next = NULL;
    return head;
  }
};


int main() {
  Solution sol;

  ListNode* head = new ListNode(1);
  ListNode* n = new ListNode(2);
  ListNode* nn = new ListNode(2);
  ListNode* nnn = new ListNode(3);
  ListNode* nnnn = new ListNode(3);
  ListNode* nnnnn = new ListNode(4);

  head->next = n;
  head->next->next = nn;
  head->next->next->next = nnn;
  head->next->next->next->next = nnnn;
  head->next->next->next->next->next = nnnnn;

  printf("Before:\n");
  head->printAll();

  sol.deleteDuplicates(head);

  printf("After:\n");
  head->printAll();
}
